e2store/src/era.rs

```diff
@@ -80,6 +80,25 @@ impl Era {
         Ok(era_state.state)
     }
 
+    /// Iterate over beacon blocks.
+    pub fn iter_blocks(raw_era1: Vec<u8>) -> impl Iterator<Item = CompressedSignedBeaconBlock> {
+        let file = E2StoreMemory::deserialize(&raw_era1).expect("invalid era1 file");
+        let entries_length = file.entries.len();
+        let block_index = SlotIndexBlockEntry::try_from(&file.entries[entries_length - 2])
+            .expect("missing block index entry")
+            .slot_index;
+
+        let mut next_slot = block_index.starting_slot;
```

Comment:
This logic can only break during the transition to the next fork. This is a best effort approach. Currently it should work for the 3 forks we are supporting, but essentially because we don't know what the next epoch is we have to guess, every time a new fork happens we just have to configure it to work, as era files shouldn't change I would think at least.

> What happens if the slot that is right before the new fork is skipped?

This shouldn't matter as long as we fix it within a 6 months window of the fork activating. The reason being is we can get the last 6 months of blocks from the CL itself.

So even if we get it wrong when a new fork is launched, we just have to fix it within 6 months so not the biggest deal

Reply:
Now that I understand the structure of the era file better, there is proper way of doing this, by using information from the `SlotIndexBlockEntry`. The `indices` field can tell us which slots are present and which ones are not (value 0 means it's absent).

I think something like this would work (but it would be good to be tested first):
```rust
block_index.indices
    .iter()
    .enumerate()
    .filter_map(|(i, index)| if index != 0 { Some(block_index.starting_slot + i) } else { None })
    .zip_eq(file.entries)  // zip_eq from itertools crate
    .map(|(slot, entry)| {
        let fork = get_beacon_fork(slot);
        CompressedSignedBeaconBlock::try_from(&entry, fork)
    })
    .collect::<Result<Vec<_>, _>>()
```

Reply:
The implementation I did is a little different then the suggestion, but it follows the same idea